### PR TITLE
Field types logic

### PR DESCRIPTION
**Pull request by ubeF**
Added a handler to field which, reads a function from field_type and executes it on players, when they land on it (or pass by it), this will be used to implement:
- The function of field_types
- Duels between players landing on the same field

It's a bit hacky, but it works

### CODE REVIEW

**Comment by hfxbse** [comment]
I will take a look at it when f26dc37dda57e58c33684b37eacd867f2df4796b and f870b6ef1a297a9705ea4a2f6d466e5300f2db09 get merged again, as those commits had issues which had not been addressed before.

You could also drop those commits from this branch if your changes don't rely on it.

**Comment by hfxbse** [comment]
It might be worth considerating creating two separate groups for the drive-by events and the turn-end events. You can then later get a [list of nodes in each group](https://docs.godotengine.org/en/stable/tutorials/scripting/groups.html#using-code). Sound promising, I might use groups for the map paths as well if I get around.

To be safe you should also [check if the desired method exists on the node](https://docs.godotengine.org/en/stable/classes/class_object.html#class-object-method-has-method) beforehand imo.

**Comment by ubeF** [comment]
> It might be worth considerating creating two separate groups for the drive-by events and the turn-end events. You can then later get a [list of nodes in each group](https://docs.godotengine.org/en/stable/tutorials/scripting/groups.html#using-code). Sound promising, I might use groups for the map paths as well if I get around.
 
Probably not worth it. Patent_office is the only driveby-event added through the map itself. The only other one would probably be the start-field, which would have a field-event AND a driveby-event (this will have to be handled completely separatly). So two separate lists wont help.

File: map/map.tscn
Comment (hfxbse) [inline map/map.tscn:1]: Only the generated IDs changed if I'm not mistaken. I do not know why it happens, but I feel weird committing it ever time it does. You can revert the file back to the version in the main branch or I'll will do it for you before merging :)

File: player/player.gd [hunk 1]
```diff
@@ -44,19 +45,29 @@ var last_transversed: Field:
...
 
 func move(distance: int):
+	self.distance = distance
 	last_location = current_location
 	moves = distance
 	await step()
 
 
 func step():
+	if moves != distance:
```
Comment (hfxbse) [inline player/player.gd]: Could the if statement and the variable `distance` be eliminated by calling the drive-by events just before `await step()` on line 74?

**Review by hfxbse — COMMENTED** [review]
I am a little confused why a drive-by event is called even when reaching the final destination unless it is a duel. Seems intentional though.

 Other than that I got no real concern, but I cannot decide if it should be `driveby` or `drive_by` as it is a hyphen normally haha

File: player/player.gd (hunk 1 above)
Comment (ubeF) [reply player/player.gd]: Yes, should have thought about that

File: map/map.tscn
Comment (ubeF) [reply map/map.tscn:1]: Feel free to revert them, my machine does this every time i launch godot

**Comment by ubeF** [comment]
> I am a little confused why a drive-by event is called even when reaching the final destination unless it is a duel. Seems intentional though.

It's so that the drive-by duel doesn't execute additionally to the normal duel. 

Since the patent_office is only defined as a drive-by event, but we also want to execute it  when we end the turn on that field, we have to handle them separately

**Comment by ubeF** [comment]
> > I am a little confused why a drive-by event is called even when reaching the final destination unless it is a duel. Seems intentional though.
> 
> It's so that the drive-by duel doesn't execute additionally to the normal duel.
> 
> Since the patent_office is only defined as a drive-by event, but we also want to execute it when we end the turn on that field, we have to handle them separately

Found a more elegant solution

**Review by hfxbse — APPROVED** [review]
Still a bit hacky, but it should do the job. You can merge it yourself when you feel happy with it :)
